store/cache.go

```diff
@@ -0,0 +1,169 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package store
+
+import (
+	"crypto"
+	"encoding/hex"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+// downloadCache is the interface that a store download cache must provide
+type downloadCache interface {
+	// Get gets the given sha3 content and puts it into targetPath
+	Get(sha3_384, targetPath string) error
+	// Put adds a new file to the cache
+	Put(sourcePath string) error
+	// Lookup checks if the given sha3 content is available in the cache
+	Lookup(sha3_384 string) bool
+}
+
+// nullCache is cache that does not cache
+type nullCache struct{}
+
+func (cm *nullCache) Get(sha3_384, targetPath string) error {
+	return fmt.Errorf("cannot get items from the nullCache")
+}
+func (cm *nullCache) Put(sourcePath string) error { return nil }
+func (cm *nullCache) Lookup(sha3_384 string) bool { return false }
+
+// changesByReverseMtime sorts by the mtime of files
+type changesByReverseMtime []os.FileInfo
+
+func (s changesByReverseMtime) Len() int           { return len(s) }
+func (s changesByReverseMtime) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+func (s changesByReverseMtime) Less(i, j int) bool { return s[i].ModTime().After(s[j].ModTime()) }
+
+// cacheManager implements a downloadCache via content based hard linking
+type CacheManager struct {
+	cacheDir string
+	maxItems int
+}
+
+// NewCacheManager returns a new CacheManager with the given cacheDir
+// and the given maximum amount of items. The idea behind it is the
+// following algorithm:
+//
+// 1. When starting a download, check if it exists in $cacheDir
+// 2. If found, update its mtime, hardlink into target location, and
+//    return success
+// 3. If not found, download the snap
+// 4. On success, hardlink into $cacheDir/<digest>
+// 5. If cache dir has more than maxItems entries, remove oldest mtimes
+//    until it has maxItems
+//
+// The caching part is done here, the downloading happens in the store.go
+// code.
+func NewCacheManager(cacheDir string, maxItems int) *CacheManager {
+	return &CacheManager{
+		cacheDir: cacheDir,
+		maxItems: maxItems,
+	}
+}
+
+// Get gets the given sha3 content and puts it into targetPath
+func (cm *CacheManager) Get(sha3_384, targetPath string) error {
+	if !cm.Lookup(sha3_384) {
+		return fmt.Errorf("cannot find %q in %q", sha3_384, cm.cacheDir)
+	}
+	if err := os.Link(cm.path(sha3_384), targetPath); err != nil {
+		return err
+	}
+	return os.Chtimes(targetPath, time.Now(), time.Now())
+}
+
+// Put adds a new file to the cache
+func (cm *CacheManager) Put(sourcePath string) error {
+	// happens on e.g. `snap download` which runs as the user
```

Comment:
I wonder if we should have some sort of locking here. Ideally we'd do a flock on file but it could be a simple mutex as long as we keep one instance of a cache in the snapd process.

Reply:
> I wonder if we should have some sort of locking here. Ideally we'd do a flock on file but it could be a simple mutex as long as we keep one instance of a cache in the snapd process.

Even better would be making sure that the operations below work correctly in the face of concurrency.